[WIP] ABFE protocol

IAlibay opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
see https://regro.github.io/rever-docs/news.html for details on how to add news entry (you do not need to run the rever command)
-->

Checklist
* [ ] Added a ``news`` entry

## Developers certificate of origin
- [x] I certify that this contribution is covered by the MIT License [here](https://github.com/OpenFreeEnergy/openfe/blob/main/LICENSE) and the **Developer Certificate of Origin** at <https://developercertificate.org/>.

IAlibay commented on openfe/protocols/openmm_utils/omm_settings.py:
```diff
@@ -660,3 +660,98 @@ class Config:
+    K_thetaA: FloatQuantity['kilocalorie_per_mole / radians ** 2'] = 20 * unit.kilocalorie_per_mole / unit.radians**2
+    """
+    The spring constant for
+      angle(host_atoms[1], host_atoms[2], guest_atoms[2])
```
```suggestion
      angle(host_atoms[1], host_atoms[2], guest_atoms[0])
```

IAlibay commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1182 @@
+                    0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
+                    0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1.0],
+                lambda_restraints=[
+                    0.0, 0.2, 0.4, 0.6, 0.8, 1.0,
```
This is wrong because it gets flipped later, i.e. `1.0 * E` means "full restraint".

IAlibay replied:
TODO: check how SepTop is doing this.

hannahbaumann replied:
Yes, this is matching now!

IAlibay commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1182 @@
+            )
+
+        if self.verbose:
+            self.logger.info(f"restraint geometry is: {rest_geom}")
```
This is a hold over from old code, probably either way would work for me.

IAlibay replied:
This is a hold over from old code, probably either way would work for me.

jthorton commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1403 @@
+        solv_dG = _get_average(individual_estimates["solvent"])
+        standard_state_dG = _get_average(individual_estimates["standard_state"])
+
+        return -complex_dG + solv_dG + standard_state_dG
```
I agree that this should be the correct way to apply the corrections!

hannahbaumann replied:
Ok, made the change!

IAlibay commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1403 @@
+        restraint : BoreschRestraint
+          A factory class for generating Boresch restraints in OpenMM.
+        """
+        frc_const = (settings.K_thetaA + settings.K_thetaB) / 2
```
Change this to be max(a, b)

hannahbaumann commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1329 @@
+    FlatBottomRestraintSettings,
+    IntegratorSettings,
+    LambdaSettings,
+    ABFEPreEquilOuputSettings,
```
```suggestion
    ABFEPreEquilOutputSettings,
```

hannahbaumann commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1329 @@
+            * engine_settings : OpenMMEngineSettings
+            * integrator_settings : IntegratorSettings
+            * equil_simulation_settings : MDSimulationSettings
+            * equil_output_settings : ABFEPreEquilOuputSettings
```
```suggestion
            * equil_output_settings : ABFEPreEquilOutputSettings
```

jthorton commented on devtools/data/gen_serialized_results.py:
```diff
@@ -73,22 +97,27 @@ def execute_and_serialize(dag, protocol, simname):
 
-    with gzip.open(f"{simname}_json_results.gz", 'wt') as zipfile:
-        json.dump(outdict, zipfile, cls=JSON_HANDLER.encoder)
+        with gzip.open(f"{simname}_json_results.gz", 'wt') as zipfile:
```
Why not use the new `to_json` in gufe is this due to CLI compatibility?

IAlibay commented on openfe/protocols/openmm_md/plain_md_methods.py:
```diff
@@ -464,6 +466,13 @@ def _run_MD(simulation: openmm.app.Simulation,
         t0 = time.time()
         simulation.step(prod_steps)
         t1 = time.time()
+        # Write out the final frame --temp debug
```
TODO work out if this was still needed.

IAlibay commented on openfe/protocols/openmm_afe/equil_afe_settings.py:
```diff
@@ -51,62 +56,117 @@ class LambdaSettings(SettingsBaseModel):
     Length of this list needs to match length of lambda_elec and lambda_restraints.
     """
+    # fmt: off
     lambda_restraints: list[float] = [
```
@hannahbaumann @jthorton please double check this - I just reversed the restraint lambdas so that it now matches everything else where 1 is non-interacting and 0 is fully interacting.

hannahbaumann replied:
This looks good, but in SepTop we're not reversing the `lambda_restraints`, should we then change it there as well?

IAlibay replied:
Are we still reversing the lambda restraints? I thought I had undone that too?

hannahbaumann commented on devtools/data/gen_serialized_results.py:
```diff
@@ -64,7 +81,12 @@ def get_hif2a_inputs():
+    dag,
+    protocol,
+    simname,
+    new_serialization: bool = False
```
Potentially add a doc string for the arguments, though not strictly needed as only devs will use this.

IAlibay replied:
Added to this method (since it's confusing and we're adding things), but not all the other rmethods since that's a lot (can do in a separate issue if you want).

hannahbaumann commented on devtools/data/gen_serialized_results.py:
```diff
@@ -109,6 +136,54 @@ def generate_md_json(smc):
+    settings.solvent_equil_simulation_settings.production_length = 10 * unit.picosecond
+    settings.solvent_simulation_settings.equilibration_length = 100 * unit.picosecond
+    settings.solvent_simulation_settings.production_length = 500 * unit.picosecond
+    settings.solvent_simulation_settings.early_termination_target_error = 0.12 * unit.kilocalorie_per_mole
```
Just out of curiosity, this shouldn't lead to early termination here since there is a minimum simulation length of 500(?) ps or so , right?

hannahbaumann replied:
I think I was mostly referring to this:
`real_time_analysis_minimum_time: PicosecondQuantity = 500.0 * unit.picosecond`, which is related to `real_time_analysis_minimum_iterations`;  that it shouldn't try early termination before 500ps if I understand it correctly?

hannahbaumann commented on openfe/protocols/openmm_afe/equil_afe_settings.py:
```diff
@@ -217,3 +278,115 @@ def must_be_positive(cls, v):
+    solvent_lambda_settings: LambdaSettings
+    """
+    Settings for controlling the solvent transformation leg
+    lambda schedule for the different components (vdw, elec, restraints).
```
Do we want to add something here that in the solvent, no restraints are implemented/applied or is that better elsewhere?

IAlibay replied:
done

hannahbaumann commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1330 @@
+        return {
+            "solvent": solv_dGs,
+            "complex": complex_dGs,
+            "standard_state": correction_dGs,
```
Maybe `"standard_state_correction"` would be more specific, though I realized we also don't do that in SepTop, maybe something to eb changed later?

hannahbaumann commented on openfe/tests/protocols/openmm_abfe/test_abfe_protocol.py:
```diff
@@ -0,0 +1,537 @@
+        else:
+            assert len(system.getForces()) == 9
+
+        # Check that the nonbonded force
```
incomplete doc string

IAlibay commented on openfe/protocols/openmm_afe/base.py:
```diff
@@ -939,6 +998,8 @@ def _run_simulation(
           The reporter associated with the sampler.
         settings : dict[str, SettingsBaseModel]
           The dictionary of settings for the protocol.
+        standard_state_corr : Optional[openff.units.Quantity]
```
We're now meant to be a bit more explicit with our units and Quantities since they clash in name between OpenMM and OpenFF - I'm not consistent, but this should be the preferred style.

IAlibay commented:
Thanks for the review @hannahbaumann !

> charged ligand test

Odd, I could have sworn I had added one, but now I'm wondering if I'm getting mixed with the ahfe test I added in a separate PR.

IAlibay commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1330 @@
+        standard_state_err = _get_stdev(individual_estimates["standard_state"])
+
+        # return the combined error
+        return np.sqrt(complex_err**2 + solv_err**2 + standard_state_err**2)
```
Done here, will follow-up in SepTop. Please double check though!

hannahbaumann commented on openfe/protocols/openmm_afe/equil_binding_afe_method.py:
```diff
@@ -0,0 +1,1377 @@
+        individual_estimates = self.get_individual_estimates()
+
+        complex_err = _get_stdev(
+            self._add_complex_standard_state_corr(
```
This looks great, thanks!

hannahbaumann reviewed (approved):
@IAlibay I'll approve the PR, pending the lambda restraints discussion on Monday!

hannahbaumann commented on openfe/protocols/openmm_afe/equil_afe_settings.py:
```diff
@@ -51,62 +56,126 @@ class LambdaSettings(SettingsBaseModel):
+        if not monotonic:
+            errmsg = (
+                "The lambda schedule is not monotonically increasing, "
+                f"got the follow schedule {v}."
```
Maybe something like this?
```suggestion
                f"got the following schedule: {v}."
```

hannahbaumann reviewed (approved):
Thanks @IAlibay , the restraints settings look good to me! just left a spelling suggestion and a question regarding some comments, but will already approve since those are not that important.

IAlibay commented on openfe/protocols/openmm_afe/equil_afe_settings.py:
```diff
@@ -51,62 +56,126 @@ class LambdaSettings(SettingsBaseModel):
         0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
         0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
     ]
+    # fmt: on
```
This is for the auto formatting tool `black` - it prevents the tool from formatting the lists and making them go one number per line. I prefer this arrangement because it lets you easily see how the schedule changes.
